plugins/Goals/Commands/CalculateConversionPages.php

```diff
@@ -0,0 +1,339 @@
+        }
+
+        $sites = explode(',',  $idSite);
+        foreach ($sites as $id) {
+            // validate the site ID
+            try {
+                new Site($id);
+            } catch (\Exception $ex) {
+                throw new \InvalidArgumentException("Invalid site ID: $id", $code = 0, $ex);
+            }
+        }
+
+        return $idSite;
+    }
+
+    /**
+     * Validate the goals parameter
+     *
+     * @return string|null
+     */
+    private function getGoalsToCalculate(): ?string
+    {
+        $idGoal = $this->getInput()->getOption('idgoal');
+
+        if(is_null($idGoal)) {
+            return null;
+        }
+
+        // Only allow the goals parameter to be used if a single site is specified
+        $idSite = $this->getInput()->getOption('idsite');
+        if (!is_numeric($idSite) || strpos($idSite, ',') !== false) {
+            throw new \InvalidArgumentException("The goals parameter can only be used when a single website is specified using the idsite parameter", $code = 0);
+        }
+
+        $goals = explode(',',  $idGoal);
+        $goalsModel = new GoalsModel();
+
+        foreach ($goals as $id) {
+            // validate the goal id
+            if (!$goalsModel->doesGoalExist($id, $idSite) && $id != GoalManager::IDGOAL_ORDER) {
+                throw new \InvalidArgumentException("Invalid goal ID: $id", $code = 0);
+            }
+        }
+
+        return $idGoal;
+    }
+
+    /**
+     * Generates the queries to calculate the 'pages before' metric for conversions within the specified date range,
+     * belonging to the specified site (if any) and specific goals (only if a single site is specified).
+     *
+     * @param string|null   $startDatetime A datetime string. Visits that occur at this time or after are deleted. If not supplied,
+     *                                     visits from the beginning of time are deleted.
+     * @param string|null   $endDatetime A datetime string. Visits that occur before this time are deleted. If not supplied,
+     *                                   visits from the end of time are deleted.
+     * @param int|null      $lastN  Calculate the last N conversions, should not be used with a date range
+     * @param string|null   $idSite The site for which to calculate, or list of comma separated sites
+     * @param string|null   $idGoal The goal for which to calculate, or list of comma separated idgoals (only if single site)
+     * @param bool          $forceRecalc If enabled then values will be recalculated for conversions that already have a
+     *                                   'pages before' value. By default only conversions with a null value will be calculated.
+     *
+     * @return array An array of queries and bind arrays   [['sql' => QUERY1, 'bind' => [PARAM1 => VALUE], ...]
+     */
+    public static function getQueries(?string $startDatetime, ?string $endDatetime, ?int $lastN = null, ?string $idSite = null,
```

Comment:
Guess there is no need to have that method public. 
```suggestion
    private static function getQueries(?string $startDatetime, ?string $endDatetime, ?int $lastN = null, ?string $idSite = null,
```